plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -354,6 +357,55 @@ public boolean isSourceModified() {
         return sourceModified;
     }
 
+    static Object deepCopy(Object value) {
+        if (value instanceof Map) {
+            @SuppressWarnings("unchecked")
+            Map<Object, Object> mapValue = (Map<Object, Object>) value;
+            return deepCopyMap(mapValue);
+        } else if (value instanceof List) {
+            @SuppressWarnings("unchecked")
+            List<Object> listValue = (List<Object>) value;
+            return deepCopyList(listValue);
+        } else {
+            return value;
+        }
+    }
+
+    private static Map deepCopyMap(Map<Object, Object> value) {
+        Map<Object, Object> copy = new HashMap<>(value.size());
+        for (Map.Entry<Object, Object> entry : value.entrySet()) {
+            Object entryValue = entry.getValue();
+            if (entryValue instanceof Map) {
+                @SuppressWarnings("unchecked")
+                Map<Object, Object> entryMapValue = (Map<Object, Object>) entryValue;
+                entryValue = deepCopyMap(entryMapValue);
+            } else if (entryValue instanceof List) {
+                @SuppressWarnings("unchecked")
+                List<Object> entryListValue = (List<Object>) entryValue;
+                entryValue = deepCopyList(entryListValue);
+            }
+            copy.put(entry.getKey(), entryValue);
+        }
+        return copy;
+    }
+
+    private static List deepCopyList(List<Object> values) {
+        List<Object> copy = new ArrayList<>(values.size());
+        for (Object value : values) {
+            if (value instanceof Map) {
+                @SuppressWarnings("unchecked")
+                Map<Object, Object> mapValue = (Map<Object, Object>) value;
+                value = deepCopyMap(mapValue);
+            } else if (value instanceof List) {
+                @SuppressWarnings("unchecked")
+                List<Object> listValue = (List<Object>) value;
+                value = deepCopyList(listValue);
+            }
```

Comment:
shall we share the two ifs in a single method called deepCopyObject(Object object) ?